src/NuGet.Services.AzureSearch/BatchPusher.cs

```diff
@@ -0,0 +1,257 @@
+            }
+
+            _versionListDataResults.Add(packageId, indexActions.VersionListDataResult);
+        }
+
+        public async Task PushFullBatchesAsync()
+        {
+            await PushBatchesAsync(onlyFull: true);
+        }
+
+        public async Task FinishAsync()
+        {
+            await PushBatchesAsync(onlyFull: false);
+        }
+
+        private async Task PushBatchesAsync(bool onlyFull)
+        {
+            await PushBatchesAsync(_hijackIndexClient, _hijackActions, onlyFull);
+            await PushBatchesAsync(_searchIndexClient, _searchActions, onlyFull);
+        }
+
+        private async Task PushBatchesAsync(
+            ISearchIndexClientWrapper indexClient,
+            Queue<IdAndValue<IndexAction<KeyedDocument>>> actions,
+            bool onlyFull)
+        {
+            while ((onlyFull && actions.Count >= _options.Value.AzureSearchBatchSize)
+                || (!onlyFull && actions.Count > 0))
+            {
+                var allFinished = new List<IdAndValue<ResultAndAccessCondition<VersionListData>>>();
+                var batch = new List<IndexAction<KeyedDocument>>();
+
+                while (batch.Count < _options.Value.AzureSearchBatchSize && actions.Count > 0)
+                {
+                    var idAndValue = DequeueAndDecrement(actions, out int newCount);
+                    batch.Add(idAndValue.Value);
+
+                    if (newCount == 0)
+                    {
+                        allFinished.Add(NewIdAndValue(idAndValue.Id, _versionListDataResults[idAndValue.Id]));
+                        Guard.Assert(_versionListDataResults.Remove(idAndValue.Id), "The version list data result should have existed.");
+                    }
+                }
+
+                await IndexAsync(indexClient, batch);
+
+                if (allFinished.Any())
+                {
+                    var versionListIdSample = allFinished
+                       .OrderByDescending(x => x.Value.Result.VersionProperties.Count(v => v.Value.Listed))
+                       .Select(x => x.Id)
+                       .Take(5)
+                       .ToArray();
+                    _logger.LogInformation(
+                        "Updating {VersionListCount} version lists, including {IdSample}.",
+                        allFinished.Count,
+                        versionListIdSample);
+                    var stopwatch = Stopwatch.StartNew();
+
+                    foreach (var finished in allFinished)
+                    {
+                        _logger.LogDebug("Updating version list for package ID {PackageId}.", finished.Id);
+                        await _versionListDataClient.ReplaceAsync(
+                            finished.Id,
+                            finished.Value.Result,
+                            finished.Value.AccessCondition);
+                    }
+
+                    stopwatch.Stop();
+                    _logger.LogInformation(
```

Comment:
Could we make this a metric?

Reply:
Tracked here: https://github.com/NuGet/NuGetGallery/issues/6664